x/dsmr/storage.go

```diff
@@ -24,33 +27,91 @@ const (
 	minSlotByte byte = 0x00
 
 	chunkKeySize = 1 + consts.Uint64Len + ids.IDLen
+
+	maxChunkStorageAheadDuration = int64(120 * time.Second)
 )
 
 var minSlotKey []byte = []byte{metadataByte, minSlotByte}
 
+var (
+	ErrChunkProducerNotValidator = errors.New("chunk producer is not in the validator set")
+	ErrChunkTooOld               = errors.New("chunk is too old")
+	ErrChunkTooFarAhead          = errors.New("chunk is too far ahead")
+	ErrInvalidChunkCertificate   = errors.New("invalid chunk certificate")
+)
+
 type Verifier[T Tx] interface {
 	Verify(chunk Chunk[T]) error
+	SetMin(min int64)
+	VerifyCertificate(ctx context.Context, chunkCert *ChunkCertificate) error
 }
 
 var _ Verifier[Tx] = (*ChunkVerifier[Tx])(nil)
 
 type ChunkVerifier[T Tx] struct {
-	networkID uint32
-	chainID   ids.ID
+	networkID  uint32
+	chainID    ids.ID
+	chainState pChain
+	min        int64
+	quorumNum  uint64
+	quorumDen  uint64
+}
+
+func NewChunkVerifier[T Tx](networkID uint32, chainID ids.ID, chainState pChain, quorumNum, quorumDen uint64) *ChunkVerifier[T] {
+	verifier := &ChunkVerifier[T]{
+		networkID:  networkID,
+		chainID:    chainID,
+		chainState: chainState,
+		quorumNum:  quorumNum,
+		quorumDen:  quorumDen,
+	}
+	return verifier
+}
+
+func (c *ChunkVerifier[T]) SetMin(min int64) {
+	c.min = min
 }
 
 func (c ChunkVerifier[T]) Verify(chunk Chunk[T]) error {
-	// TODO:
 	// check if the expiry of this chunk isn't in the past or too far into the future.
+	if chunk.Expiry < c.min {
+		return fmt.Errorf("%w: chunk expiry %d, minimum expiry %d", ErrChunkTooOld, chunk.Expiry, c.min)
+	}
+	if chunk.Expiry >= c.min+maxChunkStorageAheadDuration {
+		return fmt.Errorf("%w: chunk expiry %d, minimum expiry %d", ErrChunkTooFarAhead, chunk.Expiry, c.min)
+	}
 
-	// TODO:
 	// check if the producer was expected to produce this chunk.
+	validatorSet, err := c.chainState.GetValidatorSet(context.Background(), 0, ids.Empty)
```

Comment:
We should use a `subnetID` param here rather than fetching with subnetID `ids.Empty`